tawazi/dag.py

```diff
@@ -33,59 +31,54 @@ class DAG(Generic[P, RVDAG]):
         * Parallelization constraint of each ExecNode (is_sequential attribute)
     """
 
-    # TODO: transform into basemodel to do validation
     def __init__(
         self,
-        exec_nodes: List[ExecNode],
+        exec_nodes: Dict[Identifier, ExecNode],
+        input_uxns: List[UsageExecNode],
+        return_uxns: ReturnUXNsType,
         max_concurrency: int = 1,
         behavior: ErrorStrategy = ErrorStrategy.strict,
     ):
         """Constructor of the DAG. Should not be called directly. Instead use the `dag` decorator.
 
         Args:
             exec_nodes: all the ExecNodes
+            input_uxns: all the input UsageExecNodes
+            return_uxns: the return UsageExecNodes. These can be of various types: None, a single value, tuple, list, dict.
             max_concurrency: the maximal number of threads running in parallel
             behavior: specify the behavior if an ExecNode raises an Error. Three option are currently supported:
                 1. DAG.STRICT: stop the execution of all the DAG
                 2. DAG.ALL_CHILDREN: do not execute all children ExecNodes, and continue execution of the DAG
                 2. DAG.PERMISSIVE: continue execution of the DAG and ignore the error
         """
-        self.graph_ids = DiGraphEx()
+        self.max_concurrency = max_concurrency
+        self.behavior = behavior
+        self.return_uxns = return_uxns
+        self.input_uxns = input_uxns
 
         # ExecNodes can be shared between Graphs, their call signatures might also be different
         # NOTE: maybe this should be transformed into a property because there is a deepcopy for node_dict...
         #  this means that there are different ExecNodes that are hanging arround in the same instance of the DAG
-        self.exec_nodes = exec_nodes
```

Comment:
removing the exec_nodes attribute is a huge breaking change, what's the intended replacement ?

Reply:
This is not supposed to be a breaking change because the constructor of the DAG should not be used by the users